### ISSUE TITLE
Numpy slicing causes Buffer that shares data cannot have size different than total underlying allocation

### ISSUE THREAD

**Issue by marisancans** [NVIDIA/DALI#2877]
Hello, Im trying to use fn.external_source to load numpy array (preprocessed Mel spectrogram from audio)
When I load the array and slice it, I get a wierd error im not sure how to fix. The error message is not understandable at all to me :(

```
audio = np.load(audio_chunks_path) 
audio = audio[:30]

# This does not help
audio = np.ascontiguousarray(audio)

return images, audio
```

Heres the error:
```
Exception has occurred: RuntimeError
Critical error in pipeline:
Error when executing CPU operator MakeContiguous, instance name: "__MakeContiguous___ExternalSource_1", encountered:
[/opt/dali/dali/pipeline/data/buffer.h:234] Assert on "new_num_bytes == num_bytes_ || new_num_bytes == 0" failed: Buffer that shares data cannot have size different than total underlying allocation
Stacktrace (11 entries):
[frame 0]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/backend_impl.cpython-38-x86_64-linux-gnu.so(+0x59eb6) [0x7fb9c5468eb6]
[frame 1]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/backend_impl.cpython-38-x86_64-linux-gnu.so(dali::Buffer<dali::CPUBackend>::set_type(dali::TypeInfo const&)+0x175) [0x7fb9c5476105]
[frame 2]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/libdali.so(dali::TensorVector<dali::CPUBackend>::set_type(dali::TypeInfo const&)+0x5d) [0x7fb9a004a6ed]
[frame 3]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/libdali.so(void dali::Executor<dali::AOT_WS_Policy<dali::UniformQueuePolicy>, dali::UniformQueuePolicy>::RunHelper<dali::HostWorkspace>(dali::OpNode&, dali::HostWorkspace&)+0x71c) [0x7fb9a009e99c]
[frame 4]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/libdali.so(dali::Executor<dali::AOT_WS_Policy<dali::UniformQueuePolicy>, dali::UniformQueuePolicy>::RunCPUImpl()+0x354) [0x7fb9a00a23c4]
[frame 5]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/libdali.so(dali::Executor<dali::AOT_WS_Policy<dali::UniformQueuePolicy>, dali::UniformQueuePolicy>::RunCPU()+0xe) [0x7fb9a00a342e]
[frame 6]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/libdali.so(+0xbc5ed) [0x7fb9a005b5ed]
[frame 7]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/libdali.so(+0x13a8ac) [0x7fb9a00d98ac]
[frame 8]: /home/maris/anaconda3/envs/ml2/lib/python3.8/site-packages/nvidia/dali/libdali.so(+0x8af21f) [0x7fb9a084e21f]
[frame 9]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x8164) [0x7fba7580a164]
[frame 10]: /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7fba75732def]

Current pipeline object is no longer valid.
  File "/home/maris/src/asya/asya_vbfer_v2/datasets/v2/ds_1_zuko/dataloader.py", line 67, in __iter__
    for n_data, data in enumerate(self.pipe_iter):
  File "/home/maris/src/asya/asya_vbfer_v2/training/trainer.py", line 145, in start
    for n_batch, batch in enumerate(data_loader):
  File "/home/maris/src/asya/asya_vbfer_v2/main.py", line 90, in main
    trainer.start(dataloaders)
  File "/home/maris/src/asya/asya_vbfer_v2/main.py", line 98, in <module>
    main()
```

**Comment by marisancans** [comment]
Converting the numpy array to list and then back to numpy array works, but seems idiotic and is incredibly slow

**Comment by marisancans** [comment]
Ok, I found a fix, you have to use np.uint8 format instead of np.float64
A type check somewhere in dali would be nice.

**Comment by JanuszL** [comment]
Hi @marisancans ,

We do check types in DALI.
Could you provide a minimal repro that would illustrate the problem so we can fix it?

**Comment by marisancans** [comment]
ah, sorry, for some reason I cannot reproduce it anymore, now im getting type errors like you said. I did reinstall librosa and ffmpeg, maybe that could be the reason why its working now.

**Comment by JanuszL** [comment]
@marisancans,

Do you have a different number of samples in a batch or just a different shape?
Ad. 1 is not possible, 2 is the normal way, like each of the returned JPEGS has a different length.

**Comment by marisancans** [comment]
jpegs and labels can have different sizes each time, 1 to 30, but audio will always have length of 1

What Im basically trying to do is load N images (sequences I get from fffmpeg) and then load 1 audio file (which I then will use in dali to get mel spectograms)

**Comment by JanuszL** [comment]
> What Im basically trying to do is load N images (sequences I get from fffmpeg) and then load 1 audio file (which I then will use in dali to get mel spectograms)

It doesn't work like this. DALI operates on batches of samples. In your case, you should consider each sample as 30 images and 1 audio file and then load batch_size of them. Keep in mind that image.decoder doesn't support sequences, so either you need to load a sequence of decoded images or split it into 30 separate data nodes inside DALI and decode separately (this will make DALI to instantiate 30 image decoders and may cause OOM as each instance of the mixed image decoder consumes a piece of the GPU memory).

**Comment by marisancans** [comment]
Yea, I understand, I tried to do something like you described, but I couldnt figure out a normal way to do it, it felt like I had to fight against and break how DALI is intended to be used. 
Currently I am using a loop to get 30 images x batch_size from pipeline and then when I have full batch, I pass it to the training code. It might not be optimal, but it is 2x times faster than solution without dali. Thanks!
